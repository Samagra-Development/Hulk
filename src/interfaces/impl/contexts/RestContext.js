"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class RestContext {
    constructor(host, method, protocol, path, headers) {
        this.host = host;
        this.method = method;
        this.protocol = protocol;
        this.path = path;
        this.headers = headers;
    }
    update(key, value) {
        this[key] = value;
    }
    clear(key) {
        delete this[key];
    }
    get(key) {
        return this[key];
    }
}
exports.default = RestContext;
//# sourceMappingURL=RestContext.js.map