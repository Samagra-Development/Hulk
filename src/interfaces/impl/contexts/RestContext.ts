import { Method } from "got/dist/source/as-promise/types";

export default class RestContext implements IContext {
    host: string;
    method: Method;
    protocol: string;
    path: string;
    queryParams: Map<String, String>;
    headers: Map<String, String>;

    constructor(host: string, method: Method, protocol: string, path: string, headers: Map<String, String>) {
        this.host = host;
        this.method = method;
        this.protocol = protocol;
        this.path = path;
        this.headers = headers;
    }
    update(key: string, value: any) {
        this[key] = value;
    }
    clear(key: string) {
        delete this[key];
    }

    get(key: string) {
        return this[key];
    }
}