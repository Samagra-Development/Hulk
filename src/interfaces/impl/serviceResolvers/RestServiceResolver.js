"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const got_1 = require("got");
class RestServiceResolver {
    constructor() {
    }
    resolve(context, cadence) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { protocol, host, path, method, headers } = context;
            const { retries, timeout, linearBackoff } = cadence;
            const url = `${protocol}://${host}${path}`;
            // console.log(url, method, headers, retries, linearBackoff, timeout);
            try {
                return yield got_1.default(url, {
                    method,
                    headers,
                    retry: {
                        limit: retries,
                        calculateDelay: ({ attemptCount, retryOptions }) => retryOptions.limit >= attemptCount ? linearBackoff : 0
                    },
                    timeout
                }).text();
            }
            catch (error) {
                console.error(error);
                throw new Error("ad");
            }
        });
    }
}
exports.default = RestServiceResolver;
//# sourceMappingURL=RestServiceResolver.js.map