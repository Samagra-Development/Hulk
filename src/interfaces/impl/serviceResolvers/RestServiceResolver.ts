import got from "got";

import RestContext from "../contexts/RestContext";

export default class RestServiceResolver implements IServiceResolver {
  validator: IValidator;
  constructor() {

  }
  async resolve(context: RestContext, cadence: Cadence) {
    const { protocol, host, path , method, headers } = context;
    const { retries, timeout, linearBackoff } = cadence;
    const url = `${protocol}://${host}${path}`;
    // console.log(url, method, headers, retries, linearBackoff, timeout);

    try {
      return await got(url, {
        method,
        headers,
        retry: {
          limit: retries,
          calculateDelay: ({ attemptCount, retryOptions }) => retryOptions.limit >= attemptCount ? linearBackoff: 0
        },
        timeout 
      }).text();
    } catch(error) {
      console.error(error);
      throw new Error("ad");
    }
  }

}