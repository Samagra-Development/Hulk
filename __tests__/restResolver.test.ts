import RestContext from "../src/interfaces/impl/contexts/RestContext";
import RestServiceResolver from "../src/interfaces/impl/serviceResolvers/RestServiceResolver";

describe("Rest Service Resolver", () => {
    test("Simple GET Service", async (done) => {
        const resolver = new RestServiceResolver();
        const context: RestContext = new RestContext(
            "jsonplaceholder.typicode.com",
            "GET",
            "https",
            "/todos/200",
            new Map()
        );
        const cadence = {
            retries: 1,
            timeout: 1000,
            concurrent: false,
            linearBackoff: 100
        }
        
        const results = await resolver.resolve(context, cadence).catch(err => {
            console.error(err);
            done();
        });
        expect(results).toMatchSnapshot();
        done();
    })
})