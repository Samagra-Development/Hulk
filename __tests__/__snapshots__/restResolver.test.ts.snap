// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"{
  \\"userId\\": 10,
  \\"id\\": 200,
  \\"title\\": \\"ipsam aperiam voluptates qui\\",
  \\"completed\\": false
}"
`;
